### PR TITLE
add logging to shame bot

### PR DESCRIPTION
**Pull request by kzfubar**

### CODE REVIEW

File: shame_script.py [hunk 1]
```diff
@@ -1,14 +1,19 @@
 import configparser
+import logging
 import os
 from datetime import datetime, time
-from typing import List, Union
```
Comment (rocket3989) [inline shame_script.py:4]: I upgraded to py 3.12 to get the 3.10 feature of Union types with '|'

File: shame_script.py [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
 
+import aiohttp
 import discord
-import requests
 from discord.ext import commands, tasks
+from table2ascii import Alignment, TableStyle, table2ascii
+
 from discord_signup import signup
-from table2ascii import table2ascii, TableStyle, Alignment
+from log_setup import trace_config
```
Comment (rocket3989) [inline shame_script.py:13]: turns out imports have an order they are supposed to be in: three blocks 
1. built ins
2. 3rd party
3. local modules

and then sorted alpha within the blocks

File: shame_script.py [hunk 3]
```diff
@@ -31,6 +36,23 @@
...
+class Task(TypedDict, total=False):
+    id: int
+    content: str
+    due: DueOptions
+    labels: List[str]
+
+
+class Label(TypedDict):
+    id: int
+    name: str
```
Comment (rocket3989) [inline shame_script.py:53]: defining an API like this makes static analysis way more powerful

File: shame_script.py [hunk 4]
```diff
@@ -39,69 +61,80 @@ async def safe_send(channel: discord.TextChannel, message: str) -> discord.Messa
...
+) -> List[Task] | None:
     headers = {"Authorization": f"Bearer {api_token}"}
 
     # Get all tasks
     params = {"filter": f"(today | overdue) & !@{label_name}"}
-    response = requests.get(TODOIST_API, params=params, headers=headers)
-    return response.json()
+    async with session.get(TODOIST_API, params=params, headers=headers) as response:
+        logger.debug("Request from todoist")
+        return await response.json()
```
Comment (rocket3989) [inline shame_script.py:73]: this still needs to check for 404, shouldn't assume json is always functional

**Review by rocket3989 — APPROVED** [review]
left some comments on my own changes for context

File: shame_script.py [hunk 5]
```diff
@@ -39,69 +61,80 @@ async def safe_send(channel: discord.TextChannel, message: str) -> discord.Messa
...
+    session: aiohttp.ClientSession,
+):
     headers = {
         "Authorization": f"Bearer {api_token}",
         "Content-Type": "application/json",
     }
 
     # First, we need to get the ID of the label from the label name
-    labels_response = requests.get(
+    async with session.get(
```
Comment (kzfubar) [inline shame_script.py:88]: what is the impact of async with session here?

File: shame_script.py [hunk 6]
```diff
@@ -222,8 +263,13 @@ async def fetch_and_send_tasks():
...
-    await signup(interaction, user_to_signup, bot)
+    try:
+        await signup(interaction, user_to_signup, bot)
+        logger.info("Signup successful for user: %s", user_to_signup.name)
+    except Exception:
+        logger.exception("Error during signup")
 
 
-bot.run(DISCORD_TOKEN)
+bot.run(DISCORD_TOKEN, log_handler=None)
```
Comment (kzfubar) [inline shame_script.py:275]: log_handler = None removes all logging from discord?

File: shame_script.py (hunk 5 above)
Comment (kzfubar) [reply shame_script.py:88]: Is this the aiohttp version of requests.get() ?

**Review by kzfubar — COMMENTED** [review]
seems good to me

if running the bot with no LogHandler does what i think it does (removes all logging from discord) maybe it's best to still float up error logs?

maybe it would be better to isolate moving to aiohttp in it's own feature, but i'm ok to merge as is. 

you'll need to +2 this, as this is technically my PR so i cannot approve

File: shame_script.py (hunk 5 above)
Comment (rocket3989) [reply shame_script.py:88]: > Is this the aiohttp version of requests.get() ?

yes

this should perform better than requests  as one http session is created and shared over all the interactions

File: shame_script.py (hunk 6 above)
Comment (rocket3989) [reply shame_script.py:275]: It blocks discord from creating it's own log handler, which duplicates messages and has its own formatter

**Comment by rocket3989** [comment]
+1 aiohttp migration should have been its own branch but it had a lot of overlap with desired logging changes (mainly I wanted to validate how the logged would work with it)

**Pull request merged**